openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxUserAgentFactory.java

```diff
@@ -0,0 +1,95 @@
+package com.openhtmltopdf.pdfboxout;
+
+/**
+ *  PdfBoxUserAgentFactory. Singleton.
+ */
+public class PdfBoxUserAgentFactory {
+
+    private static class LazyFactoryHolder {
+        private static final PdfBoxUserAgentFactory INSTANCE = new PdfBoxUserAgentFactory();
+    }
+
+    protected static PdfBoxUserAgentFactory instance() {
+        return LazyFactoryHolder.INSTANCE;
+    }
+
+    private static class LazyHolder {
+        private static final UserAgentFactory INSTANCE = new UserAgentFactory() {
+            public PdfBoxUserAgent getPdfBoxUserAgent(PdfBoxOutputDevice outputDevice) {
+                return new PdfBoxUserAgent(outputDevice);
+            }
+        };
+    }
+
+    /*
+     * In all versions of Java, the idiom enables a safe, highly concurrent lazy initialization with good performance.
+     */
+    private static UserAgentFactory factoryDef() {
+        return LazyHolder.INSTANCE;
+    }
+
+    private ThreadLocal<UserAgentFactory> threadLocal;
+    private UserAgentFactory global;
+
+
+    {
+        threadLocal = new ThreadLocal<UserAgentFactory>();
+    }
+
+    /**
+     * Register [global] own / custom PdfBoxUserAgentFactory
+     * @param factory - implementation of UserAgentFactory interface
+     */
+    public void registerGlobalPdfBoxUserAgentFactory(UserAgentFactory factory) {
+        this.global = factory;
+    }
+
+    /**
+     * UnRegister [global] own / custom PdfBoxUserAgentFactory
+     */
+    public void unregisterGlobalPdfBoxUserAgentFactory() {
+        this.global = null;
+    }
+
+    /**
+     * Register [threadLocal] own / custom PdfBoxUserAgentFactory
+     * @param factory - implementation of UserAgentFactory interface
+     */
+    public  void registerPdfBoxUserAgentFactory(UserAgentFactory factory) {
+        threadLocal.set(factory);
+    }
+
+    /**
+     * UnRegister [threadLocal] own / custom PdfBoxUserAgentFactory
+     */
+    public  void unregisterPdfBoxUserAgentFactory() {
+        threadLocal.remove();
```

Comment:
If you forget this, you will have a class loader leak in your servlet container on unload....

Reply:
just like with streams, if you opened it - you must close it. It is not an argument not to use them at all. Tomcat trying to reuse user threads, which could lead to problems with threadlocal ,what why tomcat developers recommends to clean them manually. Anyway tomcat was never a good choice for a production use in a serious projects.

Reply:
This library can / will be used in whatever context the users want. Also it might be used with Tomcat (or Jetty, which also reuses threads). If we already know that using ThreadLocals will cause problems if the user is not careful we should avoid it - especially in the public API. The goal should always be to design an API in such a way that the user can not shoot himself into the foot. Not everyone has the time / is willing to read the documentation carefully. Being surprised by some "out of PermGen" errors in production is not so funny - especially if you don't have any clue where this could come from, as you usually have may different libraries in use.